Mathlib/Logic/Function/Basic.lean

```diff
@@ -991,6 +991,12 @@ def Set.piecewise {α : Type u} {β : α → Sort v} (s : Set α) (f g : ∀ i,
   fun i ↦ if i ∈ s then f i else g i
 #align set.piecewise Set.piecewise
 
+/-- `s.piecewise' f g` is a variant of `s.piecewise f g` when `f` is only defined on `s` and `g`
+    is only defined on its complement. -/
+def Set.piecewise' {α : Type _} {β : α → Sort _} (s : Set α) (f : ∀ i : {i // i ∈ s}, β i.val)
+    (g : ∀ i : {i // i ∉ s}, β i.val) [∀ j, Decidable (j ∈ s)] : ∀ i, β i :=
+  fun i ↦ if h : i ∈ s then f ⟨i, h⟩ else g ⟨i, h⟩
```

Comment:
I don't think it's a good idea to add yet again a new definition for this. I would be in favor of just using `if ... then ... else` directly, especially since lemmas about `dite` already cover some of your needs. Does that sound too annoying?